### PR TITLE
Update Leadeboard to current wireframe

### PR DESCRIPTION
**Pull request by znbudhwan**

### CODE REVIEW

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+struct RankingContext: Context {
+  let rank: Int
+  let name: String
+  let dist: String
+  let isUserTeam: Bool
+}
+
+class LeaderboardCardCell: ConfigurableTableViewCell {
+  static let identifier = "LeaderboardCardCell"
+  private let cardView = CardViewV2()
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:41]: New line before this please

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  private let leadersView = LeadersView()
+  private let content = View()
+  private let leaderboardView = LeaderboardView()
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:44]: New line after this please

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  override func commonInit() {
+    super.commonInit()
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:46]: New line after this please

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      $0.height.equalTo(216)
+    }
+    content.addSubview(leaderboardView) {
+      $0.leading.trailing.bottom.equalToSuperview().inset(Style.Padding.p8)
+      $0.top.equalTo(leadersView.snp.bottom)
+    }
+    cardView.addSubview(UIView(frame: .zero)) {
+      $0.leading.trailing.bottom.equalToSuperview()
+      $0.top.equalTo(content.snp.bottom)
+    }
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:56]: Are you trying to space out the view?  Why not use UIEdgeInsets or a custom empty cell?

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      $0.height.equalTo(650)
+    }
+    content.addSubview(leadersView) {
+      $0.leading.trailing.top.equalToSuperview().inset(Style.Padding.p8)
+      $0.height.equalTo(216)
+    }
+    content.addSubview(leaderboardView) {
+      $0.leading.trailing.bottom.equalToSuperview().inset(Style.Padding.p8)
+      $0.top.equalTo(leadersView.snp.bottom)
+    }
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift]: Why the leaderboard and the leaders in a single cell?  I think that it makes sense to have two different cells with custom rendering for each of the items.

File: Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  }
+  func configure(context: LeaderboardCellContext) {
+    if context.isUserTeam {
+      rankLbl.textColor = Style.Colors.green
+      distLbl.textColor = Style.Colors.green
+      teamLbl.textColor = Style.Colors.green
+    }
+    rankLbl.text = "\(context.rank)."
+    distLbl.text = "\(context.dist) mi"
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift:94]: We should localise the distance as per the user's settings, and use localization for the spelling too

File: Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    teamLbl.text = "\(context.name)"
```
Comment (compnerd) [inline Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift]: Is `context.name` not a string?  Why the string interpolation?

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  override func commonInit() {
+    super.commonInit()
+    contentView.addSubview(cardView) {
+      $0.leading.trailing.equalToSuperview().inset(Style.Padding.p24)
+      $0.top.bottom.equalToSuperview().inset(Style.Padding.p12)
+    }
+    cardView.addSubview(content) {
+      $0.leading.trailing.equalToSuperview()
+      $0.top.equalToSuperview()
+      $0.height.equalTo(650)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift]: is there a reason for the explicit height?

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    content.addSubview(leadersView) {
+      $0.leading.trailing.top.equalToSuperview().inset(Style.Padding.p8)
+      $0.height.equalTo(216)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift]: same here about explicit height

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      $0.leading.trailing.bottom.equalToSuperview().inset(Style.Padding.p8)
+      $0.top.equalTo(leadersView.snp.bottom)
+    }
+    cardView.addSubview(UIView(frame: .zero)) {
+      $0.leading.trailing.bottom.equalToSuperview()
+      $0.top.equalTo(content.snp.bottom)
+    }
+  }
+  override func layoutSubviews() {
+    super.layoutSubviews()
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:59]: no need for this method if all you're doing it calling the super function

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+ **/
+
+import UIKit
+import SnapKit
+
+class PlaceholderView: View {
+  private let lbl: UILabel = {
+    let lbl = UILabel(typography: .bodyRegular)
+    lbl.numberOfLines = 2
+    lbl.text = "Rest of the team rankings will show here."
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:37]: localized string

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    lbl.textAlignment = .center
+    return lbl
+  }()
+  override func commonInit() {
+    super.commonInit()
+    addSubview(lbl) {
+      $0.leading.trailing.equalToSuperview().inset(Style.Padding.p32)
+      $0.top.equalToSuperview().offset(Style.Padding.p16)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:46]: missing bottom constraint

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  }
+}
+
+class HeaderView: View {
+  private let teamName: UILabel = {
+    let lbl = UILabel(typography: .bodyBold)
+    lbl.numberOfLines = 0
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:54]: `numberOfLines = 0` is  already set in the convenience init

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  private let miles: UILabel = {
+    let lbl = UILabel()
+    lbl.numberOfLines = 0
+    lbl.textAlignment = .right
+    let imageAttachment = NSTextAttachment()
+    imageAttachment.image = Assets.downArrow.image
+    imageAttachment.bounds = CGRect(
+      x: 0,
+      y: -Style.Padding.p8,
+      width: imageAttachment.image!.size.width,
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:67]: don't use force unwrap

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    mileString.append(attachmentString)
+    lbl.attributedText = mileString
+    return lbl
+  }()
+  override func commonInit() {
+    super.commonInit()
+    addSubview(teamName) {
+      $0.leading.equalToSuperview().offset(Style.Padding.p16)
+      $0.centerY.equalToSuperview()
+      $0.width.lessThanOrEqualTo(200)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:84]: no magic number please

File: Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+      string: "Miles",
+      attributes: [NSAttributedString.Key.font: UIFont.boldSystemFont(ofSize: 16)])
+    mileString.append(attachmentString)
+    lbl.attributedText = mileString
+    return lbl
+  }()
+  override func commonInit() {
+    super.commonInit()
+    addSubview(teamName) {
+      $0.leading.equalToSuperview().offset(Style.Padding.p16)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/HeaderView.swift:82]: you can use inset on match edges

File: Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+
+class CollapseCell: TableViewCell {
+  static let identifier = "CollapseCell"
+  var lbl: UILabel = UILabel(typography: .bodyRegular, color: Style.Colors.grey)
+  weak var delegate: CollapseCellDelegate?
+  override func commonInit() {
+    super.commonInit()
+    isUserInteractionEnabled = true
+    let gesture = UITapGestureRecognizer(target: self, action: #selector(collapseOrExpand))
+    addGestureRecognizer(gesture)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/LeaderboardCell.swift:110]: why add a tap gesture to a cell when TableViewController handles cells being selected? Just pass a context back via `Contextable` and have the datasource reload based on the new info

File: Steps4Impact/Leaderboard/Cells/Views/LeaderboardView.swift [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+import UIKit
+import SnapKit
+
+class LeaderboardView: View {
+  private let placeholderView: PlaceholderView = PlaceholderView(frame: .zero)
+  private let tableView: UITableView = UITableView(frame: .zero, style: .grouped)
```
Comment (samisuteria) [inline Steps4Impact/Leaderboard/Cells/Views/LeaderboardView.swift:35]: You can just use an instance of `TableViewController` and set the datasource to whatever you'd like.

Comment (samisuteria) [reply Steps4Impact/Leaderboard/Cells/Views/LeaderboardView.swift:35]: it would get rid of all the table view boilerplate code below

**Review by samisuteria — CHANGES_REQUESTED** [review]

File: Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift (hunk 1 above)
Comment (znbudhwan) [reply Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:41]: I receive Trailing Whitespace violations when I add new lines. How can I disable those?

Comment (samisuteria) [reply Steps4Impact/Leaderboard/Cells/LeaderboardCardCell.swift:41]: delete the white space :)

**Pull request closed without merging**